File: Go/Day14.go
package main

import AH "./adventhelper"

import (
	"strings"
)

func parseInput(ss []string) (map[string]int, map[string]string, string) {
	rules := make(map[string]string)
	pairs := make(map[string]int)
	final := string(AH.FinalRune(ss[0]))

	for i := 2; i < len(ss); i++ {
		ps := strings.Split(ss[i], " -> ")
		rules[ps[0]] = ps[1]
	}

	for i := 0; i < len(ss[0]) - 1; i++ {
		str := ss[0][i:i+2]
		pairs[str] += 1
	}

	return pairs, rules, final
}

func applyRules(rs map[string]string, ps map[string]int) map[string]int {
	pairs := make(map[string]int)

	for pair, count := range ps {
		r := rs[pair]

		pairs[pair[:1] + r] += count
		pairs[r + pair[1:]] += count
	}

	return pairs
}

func repApplyRules(n int, rs map[string]string, ps map[string]int) map[string]int {
	for i := 0; i < n; i++ {
		ps = applyRules(rs, ps)
	}

	return ps
}

func score(ps map[string]int, final string) int {
	max := 0
	min := 1
	min <<= 62

	charCount := make(map[string]int)
	for pair, count := range ps {
		charCount[pair[:1]] += count
	}
	charCount[final] += 1

	for _, count := range charCount {
		if count > max {
			max = count
		}
		if count < min {
			min = count
		}
	}

	return max - min
}

func main() {
	js, _ := AH.ReadStrFile("../input/input14.txt")
	pairs, rules, final := parseInput(js)

	part1 := score(repApplyRules(10, rules, pairs), final)
	part2 := score(repApplyRules(40, rules, pairs), final)

	AH.PrintSoln(14, part1, part2)

	return
}

File: C++/Day14.cpp
#include "AH.h"

namespace Day14
{

	typedef std::map<std::string, std::string> Rules;
	typedef std::map<std::string, uint64_t> PairCounts;

	Rules parseInput(const std::vector<std::string> ss)
	{
		std::map<std::string, std::string> m;
		for (size_t i = 2; i < ss.size(); ++i)
		{
			auto ps = AH::SplitOnString(ss[i], " -> ");
				m[ps[0]] = ps[1];
		}

		return m;
	}

	PairCounts stringToPairs(const std::string s)
	{
		PairCounts pc;
		for (size_t i = 0; i < s.size() - 1; ++i)
			pc[s.substr(i, 2)] += 1;

		return pc;
	}

	PairCounts applyRules(const Rules rs, const PairCounts ps)
	{
		PairCounts newPairs;
		for (auto const &[pair, count] : ps)
		{
			const auto r = rs.at(pair);

			newPairs[(pair.at(0) + r)] += count;
			newPairs[(r + pair.at(1))] += count;
		}

		return newPairs;
	}

	PairCounts repAapplyRules(const unsigned int n, const Rules rs,
		                        const PairCounts ps)
	{
		auto psTemp = ps;
		for (unsigned int i = 0; i < n; ++i)
			psTemp = applyRules(rs, psTemp);

		return psTemp;
	}

	int64_t score(const PairCounts ps, const char final)
	{
		uint64_t max = 0;
		uint64_t min = 1; (min <<= 62);

		std::map<char, uint64_t> charCount;
		for (auto const & [pair, count] : ps)
			charCount[pair.at(0)] += count;

		charCount[final] += 1;

		for (auto const & [pair, count] : charCount)
		{
			if (count > max)
				max = count;
			if (count < min)
				min = count;
		}

		return max - min;
	}

	int Run(const std::string& filename)
	{
		auto inputLines = AH::ReadTextFile(filename);
		auto pairs = stringToPairs(inputLines[0]);
		auto rules = parseInput(inputLines);
		
		const auto final = inputLines[0].back();
		const auto p1 = repAapplyRules(10, rules, pairs);
		const auto p2 = repAapplyRules(40, rules, pairs);

		AH::PrintSoln(14, score(p1, final), score(p2, final));

		return 0;
	}

}
